### DISCUSSION TITLE
Reading large values from std_logic_vectors in a testbench

### DISCUSSION THREAD

**Discussion by tutsta** [cocotb/cocotb#4290] [Q&A] [answered]
Hi Folks,
I have a 64-bit std_logic_vector in a testbench that I would like to read each clock cycle and save the values in a NumPy array. The 64-bit slv is 8 x 8-bit integers packed into the 64-bit word and I would like to know if there's a way to read slices of the 64-bit  word directly in my CoCoTB testbench, rather than having to slice inside the HDL code?
Here's a code fragmentexample, where `rd_data` is the 64-bit std_logic_vector:
```
async def capture_rd_data(dut):
    rd_data = []
    idx = 0
    await RisingEdge(dut.i1.s_clk_dsp)
    while idx < NUM_OUTPUT_FRAMES:
        rd_en = dut.i1.dsp.lane[7].inc.ram[6].buf.s_rd_en_d1.value
        if rd_en.integer == 1:
            rd_data = dut.i1.dsp.lane[7].inc.ram[6].buf.s_rd_data.value
            rd_data.append(rd_data.integer)
            idx += 1
            
    np.save("rd_data", rd_data)
```
When I load the resulting NumPy file, I find that the 64-bit integer has been converted to a float, which causes the lower bits to be lost in the conversion.
I noticed that there is a `Range` argument in the documentation but it's not clear to me how to use it in this example (or if that is what it's intended for)?
Thanks in advance!

**Comment by WernerFS** [comment]
Update:

``` Python
import numpy as np
from cocotb.types import LogicArray, Range

value = 3324679375210329505 
long_word = LogicArray(value, Range(64 - 1, "downto", 0)) # equivalent to Simhandle.value
long_word_value_int = long_word.to_unsigned()

rd_data_int64 = np.array([long_word_value], dtype=np.int64)
rd_data_float64 = np.array([long_word_value], dtype=np.float64)
rd_data = np.array([long_word_value])

np.save("rd_data_int64", rd_data_int64)
np.save("rd_data_float64", rd_data_float64)
np.save("rd_data", rd_data)

rd_data2_int64 = np.load("rd_data_int64.npy")
rd_data2_float64 = np.load("rd_data_float64.npy")
rd_data2 = np.load("rd_data.npy")

print(rd_data == rd_data2_int64)
print(rd_data == rd_data2_float64)
print(rd_data == rd_data2)
```

------------------------------------------------------------------------------------------------------------

It is hard to say without having the reading part. I made a short example showing that even when Python guesses the type it does not matter. You can anyways pass the `dtype` parameter to be more conservative. 

```Python

import numpy as np
from cocotb.types import LogicArray, Range

value = 3324679375210329505 
long_word = LogicArray(value, Range(64 - 1, "downto", 0)).to_bytes("little") # 64 bit LogicArray similar to the simHandle return object
long_word_value = int.from_bytes(long_word, byteorder= "little") # simulate the .value call

rd_data_int64 = np.array([long_word_value], dtype=np.int64)
rd_data_float64 = np.array([long_word_value], dtype=np.float64)
rd_data = np.array([long_word_value])

np.save("rd_data_int64", rd_data_int64)
np.save("rd_data_float64", rd_data_float64)
np.save("rd_data", rd_data)

rd_data2_int64 = np.load("rd_data_int64.npy")
rd_data2_float64 = np.load("rd_data_float64.npy")
rd_data2 = np.load("rd_data.npy")

print(rd_data == rd_data2_int64)
print(rd_data == rd_data2_float64)
print(rd_data == rd_data2)
```

Now back to the slicing part. One way to do it is making use of the good ol` bit masks. 

``` Python
def n_bit_mask(n: int) -> int:
    return (1 << n) - 1

lower_int = rd_data & n_bit_mask(8)
second_lowest_int = (rd_data & (n_bit_mask(8) << 8)) >> 8
# so on and so forth
```

Or, go for the `bytes` approach (only works if your data is arranged in bytes)

``` Python
value = 3324679375210329505 
long_word = (value).to_bytes(8, byteorder = "little") # Create logic array from value
list(long_word) # will give you a list of length 8 being the 8 ints that form the 64 bits word
```

For `LogicArray` you can slice as using a normal array. Better for custom length data.

``` Python
long_word = LogicArray(value, Range(64 - 1, "downto", 0))
lowest_int = long_word[7:0]  # equivalent to HDL (8 downto 0)
second_lowest_int = long_word[15:8]
```

**Reply by tutsta** [reply]
Hi Werner,

Thanks so much for such a detailed reply! The technique you propose in the first part of your reply looks very promising but, unfortunately I don't seem to be able to get it to work in the actual testbench. I now have the following code fragment:
```python
    rd_data_int64 = np.arange(NUM_SAMP, dtype=np.int64)
    idx = 0
    while idx < NUM_SAMP:
        await RisingEdge(dut.i1.s_clk_dsp)
        ram_rd_en = dut.i1.dsp.lane[7].inc.ram[6].buf.s_rd_en_d1.value
        if ram_rd_en.integer == 1:
            ram_rd_data = dut.i1.dsp.lane[7].inc.ram[6].buf.s_rd_data.value
            rd_data_int64[idx] = int.from_bytes(ram_rd_data, byteorder= "little")
            idx += 1

    np.save("sample_blk_ram_rd_data_np_int64", rd_data_int64)
```
However, when I run the testbench, it fails with the following message:
```
rd_data_int64[idx] = int.from_bytes(ram_rd_data, byteorder= "little")
OverflowError: Python int too large to convert to C long
```
Perhaps I've misunderstood your technique?

Thanks and regards,
John

**Reply by WernerFS** [reply]
Oh so it looks that this is an architecture issue. Better even, just use the native casting method.

`ram_rd_data.value.to_unsigned()`

**Reply by tutsta** [reply]
That doesn't seem to work either unfortunately...
```
rd_data_int64[idx] = ram_rd_data.to_unsigned()
AttributeError: 'BinaryValue' object has no attribute 'to_unsigned'
```
This is the testbench code that generated it...
```python
    rd_data_int64 = np.arange(NUM_SAMP, dtype=np.int64)
    idx = 0
    while idx < NUM_SAMP:
        await RisingEdge(dut.i1.s_clk_dsp)
        ram_rd_en = dut.i1.dsp.lane[7].inc.ram[6].buf.s_rd_en_d1.value
        if ram_rd_en.integer == 1:
            ram_rd_data = dut.i1.dsp.lane[7].inc.ram[6].buf.s_rd_data.value
            rd_data_int64[idx] = ram_rd_data.to_integer()
            idx += 1

    np.save("sample_blk_ram_rd_data_np_int64", rd_data_int64)
```

**Reply by WernerFS** [reply, marked as answer]
Oh I see. I missed something important in my answer. It is only applicable from `cocotb 1.8`. Before that, there were no `LogicArrays`. In your case you should be able to achieve the same result using the following:

```
rd_data_int64[idx] = ram_rd_data.integer
```

Please test it and let me know. [REF](https://docs.cocotb.org/en/stable/library_reference.html#cocotb.binary.BinaryValue)

**Reply by tutsta** [reply]
Your suggestion did the trick, thanks very much!
 With this code fragment:
 ```python
    rd_data_int = []
    rd_data_int64 = np.arange(NUM_SAMP, dtype=np.uint64)
    idx = 0
    while idx < NUM_SAMP:
        await RisingEdge(dut.i1.s_clk_dsp)
        ram_rd_en = dut.i1.dsp.lane[7].inc.ram[6].buf.s_rd_en_d1.value
        if ram_rd_en.integer == 1:
            ram_rd_data = dut.i1.dsp.lane[7].inc.ram[6].buf.s_rd_data.value
            rd_data_int.append(ram_rd_data.integer)
            rd_data_int64[idx] = ram_rd_data.integer
            idx += 1

    np.save("sample_blk_ram_rd_data_int", rd_data_int)
    np.save("sample_blk_ram_rd_data_np_int64", rd_data_int64)
```
I get these values when the NumPy files are read back into Python:
```
Actual bus value   0xe83811d3fe10030b 0x807f5780c77f229e 0x7f80807f807f7f80 0x7f4c80807f807f80
rd_data_int        0xe83811d3fe100000 0x807f5780c77f2000 0x7f80807f807f8000 0x7f4c80807f808000
rd_data_int64      0xe83811d3fe10030b 0x807f5780c77f229e 0x7f80807f807f7f80 0x7f4c80807f807f80
```
As you can see, "standard" integer assignment has resulted in the lower bits being dropped, while the uint64 has faithfully captured the full bus values. The part that seemed to make the real difference was defining a NumPy array as ```dtype = np.uint64```

Thanks again for your help and time with this Werner, very much appreciated!
John
